src/core/thread/dua_manager.cpp

```diff
@@ -184,6 +253,508 @@ otError DuaManager::Store(void)
+    if (!Get<Mle::Mle>().IsRxOnWhenIdle())
+    {
+        Get<DataPollSender>().SendFastPolls();
+    }
+
+    otLogInfoDua("Sent DUA.req for DUA %s From %s to %s", dua.ToString().AsCString(),
+                 messageInfo.GetSockAddr().ToString().AsCString(), messageInfo.GetPeerAddr().ToString().AsCString());
+
+exit:
+    if (error == OT_ERROR_NO_BUFS)
+    {
+        UpdateCheckDelay(Mle::kNoBufDelay);
+    }
+
+    if (error != OT_ERROR_NONE && message != nullptr)
+    {
+        message->Free();
+    }
+}
+
+void DuaManager::HandleDuaResponse(Coap::Message &aMessage, const Ip6::MessageInfo &aMessageInfo, otError aResult)
+{
+    OT_UNUSED_VARIABLE(aMessageInfo);
+
+    mIsDuaPending = false;
+
+    if (aResult == OT_ERROR_RESPONSE_TIMEOUT)
+    {
+        UpdateCheckDelay(Mle::KResponseTimeoutDelay);
+        ExitNow();
+    }
+
+    VerifyOrExit(aResult == OT_ERROR_NONE && aMessage.GetCode() == OT_COAP_CODE_CHANGED, OT_NOOP);
+
+    ProcessDuaResponse(aMessage);
+
+exit:
+    mRegistrationTask.Post();
+}
+
+void DuaManager::HandleDuaNotification(Coap::Message &aMessage, const Ip6::MessageInfo &aMessageInfo)
+{
+    OT_UNUSED_VARIABLE(aMessageInfo);
+
+    otLogInfoDua("Received DUA.ntf");
+
+    VerifyOrExit(aMessage.GetCode() == OT_COAP_CODE_POST, OT_NOOP);
+
+    if (aMessage.IsConfirmable())
+    {
+        if (Get<Coap::Coap>().SendEmptyAck(aMessage, aMessageInfo) == OT_ERROR_NONE)
+        {
+            otLogInfoDua("Sent DUA.ntf acknowledgment");
+        }
+    }
+
+    ProcessDuaResponse(aMessage);
+
+exit:
+    return;
+}
+
+void DuaManager::ProcessDuaResponse(Coap::Message &aMessage)
+{
+    Ip6::Address target;
+    uint8_t      status;
+
+    SuccessOrExit(Tlv::FindUint8Tlv(aMessage, ThreadTlv::kStatus, status));
+    SuccessOrExit(Tlv::FindTlv(aMessage, ThreadTlv::kTarget, &target, sizeof(target)));
+
+#if OPENTHREAD_CONFIG_DUA_ENABLE
+    if (Get<ThreadNetif>().HasUnicastAddress(target))
+    {
+        switch (static_cast<ThreadStatusTlv::DuaStatus>(status))
+        {
+        case ThreadStatusTlv::kDuaSuccess:
+            mLastRegistrationTime = TimerMilli::GetNow();
+            break;
+        case ThreadStatusTlv::kDuaReRegister:
+            mDuaState                  = kToRegister;
+            mDelay.mFields.mCheckDelay = Mle::kImmediateReRegisterDelay;
+            break;
+        case ThreadStatusTlv::kDuaInvalid:
+            // Domain Prefix might be invalid.
+            RemoveDomainUnicastAddress();
+            break;
+        case ThreadStatusTlv::kDuaDuplicate:
+            RemoveDomainUnicastAddress();
+            mDadCounter++;
+
+            if (GenerateDomainUnicastAddressIid() == OT_ERROR_NONE)
+            {
+                AddDomainUnicastAddress();
+            }
+
+            break;
+        case ThreadStatusTlv::kDuaNoResources:
+            // fall through
+        case ThreadStatusTlv::kDuaNotPrimary:
+            // fall through
+        case ThreadStatusTlv::kDuaGeneralFailure:
```

Comment:
I don't think we need the `// fall through` is there is no code between the `case` statements.
```suggestion
        case ThreadStatusTlv::kDuaNoResources:
        case ThreadStatusTlv::kDuaNotPrimary:
        case ThreadStatusTlv::kDuaGeneralFailure:
```

Reply:
I think `fall through` is more explicit? Otherwise it's not so obvious that we are falling through.

Reply:
Agree @librasungirl